src/alexandria3k/__main__.py

```diff
@@ -1,536 +1,569 @@
+    """Given the user-visible name of a facility (e.g. funder-names)
+    return the corresponding name of the module (e.g funder_names)."""
+    return facility.replace("-", "_")
+
+
+def class_name(facility):
+    """Given the user-visible name of a facility (e.g. funder-names)
+    return the corresponding name of the class (e.g FunderNames)."""
+    components = facility.split("-")
+    # Capitalize the first letter of each component and join them together.
+    return "".join(x.title() for x in components)
+
+
+def facility_modules(facility):
+    """Return a list with the module names of the available facilities"""
+    main_dir = os.path.dirname(os.path.realpath(__file__))
+    python_files = os.listdir(f"{main_dir}/{facility}")
+    # Remove trailing .py
+    return [os.path.splitext(f)[0] for f in python_files if f.endswith(".py")]
+
+
+def facility_names(facility):
+    """Return a list with the names of the available facilities.
+    (data_source or process)"""
+    # Replace _ with -
+    return [s.replace("_", "-") for s in facility_modules(facility)]
+
+
+def get_data_source_instance(args):
+    """Return a data source instance based on the specified user arguments."""
+    facility = args.data_name
+    module = module_name(f"data_sources.{facility}")
+
+    default_source = module_get_attribute(module, "DEFAULT_SOURCE")
+    data_location = args.data_location or default_source
+    if not data_location:
+        raise Alexandria3kError(
+            f"The data source {facility} requires the specification of a data location"
+        )
+
+    # The sampling expression defaults to True or it can be
+    # given via the CI. One can manipulate the sampling function
+    # by using a variable called data as the input argument
+    # of the lambda function to access the input of the callable.
+    # (e.g. For USPTO sampling the input of the callable is a tuple)
+    # pylint: disable-next=eval-used
+    sample = eval(f"lambda data: {args.sample}")
+
+    class_ = module_get_attribute(module, class_name(facility))
+    return class_(data_location, sample, args.attach_databases)
+
+
+def download(args):
+    """Download data using the specified data source."""
+
+    data_source_instance = get_data_source_instance(args)
+    data_source_instance.download(args.database, args.sql_query)
+    perf.log(
+        f"Data downloaded and saved to {data_source_instance.output_path}"
+    )
+
+
+def add_subcommand_download(subparsers):
+    """Add the arguments of the download subcommand."""
+    parser = subparsers.add_parser(
+        "download", help="Download data using the specified data source."
+    )
+    parser.set_defaults(func=download)
+    parser.add_argument(
+        "database", help="File path of the database to use for retrieving data"
+    )
+    parser.add_argument(
+        "data_name",
+        choices=facility_names("data_sources"),
+        help="Name of the data source to use",
+    )
+    parser.add_argument(

```

Comment:
This should be optional (an option) where both the database and the SQL query should be given or not given.